time/time.1
.\"
.\" Sccsid @(#)time.1	1.9 (gritter) 3/10/03
.\" Parts taken from time(1), Unix 7th edition:
.\" Copyright(C) Caldera International Inc. 2001-2002. All rights reserved.
.\"
.\" SPDX-Licence-Identifier: Caldera
.\"
.TH TIME 1 "3/10/03" "Heirloom Toolchest" "User Commands"
.SH NAME
time, ptime \- time a command
.SH SYNOPSIS
\fBtime\fR [\fB\-p\fR] \fIcommand\fR
.br
\fBptime\fR [\fB\-p\fR] \fIcommand\fR
.SH DESCRIPTION
The
given command is executed; after it is complete,
.I time
prints the elapsed time during the command, the time
spent in the system, and the time spent in execution
of the command.
Times are reported in seconds.
.PP
The times are printed on the diagnostic output stream.
.PP
.I Ptime
performs a similar task using the
.I /proc
file system,
but does not include the time values of child processes
of
.IR command .
.PP
Both
.B time
and
.B ptime
accept the following option:
.TP
.B \-p
Use the portable output format, as in:
.IP
.BI real\  "(real seconds)"
.br
.BI user\  "(user seconds)"
.br
.BI sys\  "(system seconds)"
.PP
If the specified command could not get found,
.I time
and
.I ptime
exit with a status of 127.
If the command was found,
but could not get executed,
the exit status is 126.
.SH "ENVIRONMENT VARIABLES"
.TP
.BR LANG ", " LC_ALL
See
.IR locale (7).
.TP
.B LC_NUMERIC
Sets the format of numbers printed with the
.B \-p
option.
.TP
.B PATH
Used to find the location of
.IR command .
.SH "SEE ALSO"
sh(1),
times(2),
proc(5)
.SH NOTES
.I Ptime
prints values in milliseconds,
but the practical accuracy depends on the system clock.


CHANGES
Release 230305
* getconf: Now _NPROCESSORS_ONLN and _NPROCESSORS_CONF are implemented.
* Removed the remaining licence headers, now every licence is referenced by its SPDX identifier.

Release 230228
* [NEW] readlink: New implementation, extensively based on OpenBSD's.
* readlink: It's installed at /usr/ucb per default and symbolic/hard linked to /usr/5bin. 

Release 230120
* Notes for the library directories if linking against the GNU C library.
* rm: Consistency within standards: the SUS program shouldn't have the POSIX 1003.1(2008) fix, just the main program.
* Made $(ROOT) not hardcoded anymore.
* Cleaned up more licence headers (if my estimations are correct, we've got rid of 3.255 lines!).
* Small update on README.md --- although it's now less readable on plain-text readers such as more or pg.

Release 230107
* rm: Added support for the POSIX 1003.1(2008).
* chroot: Added manual page.
* awk: Added stdint.h to awk.g.y.
* Now it has a bigger range of compatibility with GNU C library versions.

Release 221025
* Made notes clearer about building Heirloom on modern versions of the GNU C library.
* tar: Support for zstd and xz.

Release 220821
* [NEW] chroot: New implementation of chroot, based on Solaris 2.5 payloads.
* true, false: Rewritten in C.
* yes: Rewritten and adapted to ANSI C.
* "libcommon/vpfmt.c": Adapted to ANSI C.
* Now many of the utilities can be built in the musl C library and GNU C library (dynamically).

Release 070715
* cpio: A "-P" option has been introduced that allows to specify attributes
  like the file owner when creating an archive. With it, unprivileged users
  can set arbitrary attributes for files within an archive.
* diff: Made it compile on CentOS 5.
* Static linking against glibc is now possible since "optopt" is no longer
  a common symbol, and there is no conflict anymore with "getopt.o".
* cpio, tar: A compilation problem on some Linux systems has been fixed.
* ul: A compilation problem for NetBSD and other systems that use termcap
  instead of curses has been fixed (patch by Aleksey Cheusov).
* bfs: An OpenSolaris-derived manual page is now included.

Release 070227
* The directory for development utilities, by default "/usr/ccs/bin", is
  now configurable in "build/mk.config" . It is expected that the utilities
  from the "Heirloom Development Tools" package are installed in it.
* getconf: "/usr/ccs/bin" is now included in the default path as printed
  with the "PATH" argument.
* what: Has been replaced by an OpenSolaris-derived implementation.
* tsort: Is now installed in "/usr/ccs/bin".
* OpenSolaris-derived manual pages are now included for bdiff, newform,
  and fspec(5).
* psrinfo: Fixed the computation of the "on-line since" time for Linux 2.6.
* cpio: Added a "-D" option to make it usable in combination with SVR4
  packaging tools. This option is not intended for general use, and is
  thus not documented.
* The package build target "make pkg" for Solaris and Open UNIX is no
  longer supported.

Release 070115
* deroff: Now recognizes the ".do" and ".xflag" requests as well as an "-xN"
  command line option to handle Heirloom nroff/troff extensions.
* ed: Substituting only a specific occurence of the pattern, as in the
  command "s/x/y/3", now works correctly if the substitution "y" is not
  equal in length to the pattern match "x" (Derek Thomas).
* ps: Use the Linux 2.6 /proc/status "SleepAVG" field to compute the value
  for the "C" column (cpu usage for scheduling).
* priocntl: The "FF", "B", and "ISO" class names are now accepted with
  the "-e" option too.
* file: UTF-8 characters in troff input files are now recognized.
* cp: The "-s" option now displays correct sizes and related statistics
  for special files with st_size == 0.
* grep: With the "-z" option, decompressor children processes are now
  properly waited for, avoiding failures when processing many files.
* su: Works with PAM on Mac OS X now (Stefan Tramm).
* ps: Handles dead processes as on other platforms now (Stefan Tramm).
* pgrep: Ignores dead processes now (Stefan Tramm).
* spell: A value of the wrong type that was passed to scanf() could lead
  to build errors on 64-bit platforms (Jeffrey H. Johnson).
* ps, pgrep, whodo: Works on FreeBSD 6 now.
* df, find, ps: Changes have been made to build on NetBSD 3 (Aleksey Cheusov).
* Changes have been made to build on OpenBSD 3.8 (Aleksey Cheusov).
* Changes have been made to build with GNU make 3.81.

Release 060122
* The tools have been ported to Mac OS X 10.4.3 (patches by Stefan Tramm).
* The tools have been ported to DragonFlyBSD 1.3.7-DEVELOPMENT.

Release 051118
* rm: The output of the question before descending to a directory with
  the "-i" option has been fixed.
* /usr/ucb/ps: The "-a" option has been fixed; it incorrectly printed a
  list of all processes since release 050411.
* getconf: /usr/ucb is now put at the end of the Heirloom directory list
  for the PATH variable.
* The "-man" macro no longer produces cut marks when used with troff.
* tr: When evaluating '[:class:]' expressions, the last position of the
  coded character set was not included; this has been fixed.
* ps: The "-R" option (resource usage format), which had been present
  for some time, has been documented.
* RPM package builds now require the new "Heirloom Development Tools"
  package from <http://heirloom.sourceforge.net/devtools.html> to be
  installed.

Release 050815
* The "getconf" utility has been added (derived from OpenSolaris code).
* The "mesg" and "pathchk" utilities have been added.
* tr: A new implementation that is based on 6th and 7th edition code is
  now used. It stores the translation tables in arrays of variable size.
  This results in better performance, under some circumstances at the
  expense of higher memory consumption.
* printf: Handle unsigned values larger than INT_MAX correctly with "%u",
  "%o", "%x", and "%X".
* printf: Handle SVID3 "position" fields in format specifications, such

time/time.c
/*
 * time - time a command
 *
 * Gunnar Ritter, Freiburg i. Br., Germany, December 2000.
 */
/*
 * Copyright (c) 2003 Gunnar Ritter
 *
 * SPDX-Licence-Identifier: Zlib
 */

#if __GNUC__ >= 3 && __GNUC_MINOR__ >= 4 || __GNUC__ >= 4
#define	USED	__attribute__ ((used))
#elif defined __GNUC__
#define	USED	__attribute__ ((unused))
#else
#define	USED
#endif
#ifdef	PTIME
static const char sccsid[] USED = "@(#)ptime.sl	1.28 (gritter) 5/29/05";
#else
static const char sccsid[] USED = "@(#)time.sl	1.28 (gritter) 5/29/05";
#endif

#include	<sys/types.h>
#include	<sys/wait.h>
#include	<sys/times.h>
#include	<unistd.h>
#include	<signal.h>
#include	"sigset.h"
#include	<stdio.h>
#include	<stdlib.h>
#include	<string.h>
#include	<errno.h>
#include	<libgen.h>
#include	<ctype.h>
#ifdef	PTIME
#include	<limits.h>
#endif
#include	<locale.h>

static int		pflag;		/* portable format */
#undef	hz
static long		hz;		/* clock ticks per second */
static char		*progname;	/* this command's name */
static struct lconv	*localec;	/* locale information */

/*
 * perror()-alike.
 */
#ifdef	PTIME
static void
pnerror(int eno, const char *string)
{
	fprintf(stderr, "%s: %s: %s\n", progname, string, strerror(eno));
}
#endif	/* PTIME */

/*
 * Write a single output line for type msg. Val is given in clock ticks.
 */
static void
out(const char *msg, clock_t val)
{
#ifndef	PTIME
	long divider = pflag ? 100 : 10;
#else
	long divider = pflag ? 100 : 1000;
#endif
	long hours, mins, secs, fracs;

	secs = val * divider / hz;
	fracs = secs % divider;
	secs = (secs - fracs) / divider;
	if (pflag)
		fprintf(stderr, "%s %lu%s%02lu\n", msg, secs,
				localec->decimal_point, fracs);
	else {
		fprintf(stderr,"%-4s ", msg);
		if (secs > 59) {
			mins = secs / 60;
			secs %= 60;
			if (mins > 59) {
				hours = mins / 60;
				mins %= 60;
				fprintf(stderr, "%2lu:%02lu:%02lu.",
						hours, mins, secs);
			} else
				fprintf(stderr, "   %2lu:%02lu.",
						mins, secs);
		} else
			fprintf(stderr, "      %2lu.", secs);
#ifdef	PTIME
		fprintf(stderr, "%03lu\n", fracs);
#else
		fprintf(stderr, "%lu\n", fracs);
#endif
	}
}

#ifdef	PTIME
/*
 * Memory allocation with check.
 */
static void *
srealloc(void *vp, size_t nbytes)
{
	void *p;

	if ((p = (void *)realloc(vp, nbytes)) == NULL) {
		write(2, "Out of memory\n", 14);
		exit(077);
	}
	return p;
}

static void *
smalloc(size_t sz)
{
	return srealloc(NULL, sz);
}

enum	valtype {
	VT_CHAR,
	VT_INT,
	VT_UINT,
	VT_LONG,
	VT_ULONG
};

union	value {
	char			v_char;
	int			v_int;
	unsigned int		v_uint;
	long			v_long;
	unsigned long		v_ulong;
};

static union value *
getval(char **listp, enum valtype type, int separator)
{
	char	*buf;
	static union value	v;
	const char	*cp, *op;
	char	*cq, *x;

	if (**listp == '\0')
		return NULL;
	op = *listp;
	while (**listp != '\0') {
		if (separator == ' ' ? isspace(**listp) : **listp == separator)
			break;
		(*listp)++;
	}
	buf = smalloc(*listp - op + 1);
	for (cp = op, cq = buf; cp < *listp; cp++, cq++)
		*cq = *cp;
	*cq = '\0';
	if (**listp) {
		while (separator == ' ' ?
				isspace(**listp) : **listp == separator)
			(*listp)++;
	}
	switch (type) {
	case VT_CHAR:
		if (buf[1] != '\0')
			return NULL;
		v.v_char = buf[0];
		break;
	case VT_INT:
		v.v_int = strtol(buf, &x, 10);
		if (*x != '\0')
			return NULL;
		break;
	case VT_UINT:
		v.v_uint = strtoul(buf, &x, 10);
		if (*x != '\0')
			return NULL;
		break;
	case VT_LONG:
		v.v_long = strtol(buf, &x, 10);
		if (*x != '\0')
			return NULL;
		break;
	case VT_ULONG:
		v.v_ulong = strtoul(buf, &x, 10);
		if (*x != '\0')
			return NULL;
		break;
	}
	return &v;
}

static int
failed(void)
{
	fprintf(stderr, "/proc read failed\n");
	return -1;
}

#define	GETVAL(a)	if ((v = getval(&cp, (a), ' ')) == NULL) \
				return failed()

/*
 * Get process times from /proc.
 */
static int
ptimes(struct tms *tp, FILE *fp, char *stfn)
{
	static char	*buf;
	static size_t	buflen;
	union value	*v;
	char *cp, *cq, *ce;
	size_t	sz, sc;

	for (cp = buf; ;) {
		const unsigned	chunk = 32;

		if (buflen < (sz = cp - buf + chunk)) {
			sc = cp - buf;
			buf = srealloc(buf, buflen = sz);
			cp = &buf[sc];
		}
		if ((sz = fread(cp, 1, chunk, fp)) < chunk) {
			ce = &cp[sz - 1];
			break;
		}
		cp += chunk;
	}
	fclose(fp);
	if (*ce != '\n')
		failed();
	*ce-- = '\0';
	cp = buf;
	GETVAL(VT_INT);
	/* pid not used */
	if (*cp++ != '(')
		failed();
	for (cq = ce; cq >= cp && *cq != ')'; cq--);
	if (cq < cp)
		failed();
	*cq = '\0';
	cp = &cq[1];
	while (isspace(*cp))
		cp++;
	GETVAL(VT_CHAR);
	/* state not used */
	GETVAL(VT_INT);
	/* ppid not used */
	GETVAL(VT_INT);
	/* pgrp unused */
	GETVAL(VT_INT);
	/* session unused */
	GETVAL(VT_INT);
	/* nr not used */
	GETVAL(VT_INT);
	/* tty_pgrp unused */
	GETVAL(VT_ULONG);
	/* flags unused */
	GETVAL(VT_ULONG);
	/* min_flt unused */
	GETVAL(VT_ULONG);
	/* cmin_flt unused */
	GETVAL(VT_ULONG);
	/* maj_flt unused */
	GETVAL(VT_ULONG);
	/* cmaj_flut unused */
	GETVAL(VT_ULONG);
	tp->tms_cutime = v->v_ulong;
	GETVAL(VT_ULONG);
	tp->tms_cstime = v->v_ulong;
	return 0;
}
#endif	/* PTIME */

/*
 * In child process: Execute the command or set error status.
 */
static void
child(char **av)
{
	int err;

	execvp(av[0], av);
	err = errno;
	fprintf(stderr, "%s: %s\n", strerror(err), av[0]);
	_exit(err == ENOENT ? 0177 : 0176);
}

/*
 * Time a command.
 */
static int
timecmd(char **av)
{
	void (*oldint)(int), (*oldquit)(int);
	struct tms tp;
#ifdef	PTIME
	char pdir[_POSIX_PATH_MAX];
	FILE *fp;
	char *stfn = "stat";
#endif
	clock_t t1, t2;
	pid_t pid;
	int status;

	oldint = sigset(SIGINT, SIG_IGN);
	oldquit = sigset(SIGQUIT, SIG_IGN);
	t1 = times(&tp);
	switch (pid = fork()) {
	case 0:
		sigset(SIGINT, oldint);
		sigset(SIGQUIT, oldquit);
		child(av);
		/*NOTREACHED*/
	case -1:
		fprintf(stderr, "%s: cannot fork -- try again.\n", progname);
		return 1;
	}
#ifdef	PTIME
	/*
	 * Changing to the child's /proc entry will keep it in zombie status
	 * even after it has been waited for.
	 */
	snprintf(pdir, sizeof pdir, "/proc/%lu", (long)pid);
	if (chdir(pdir) < 0) {
		pnerror(errno, pdir);
		return 1;
	}
	/*
	 * Starting with Linux 2.4.5, the stat file has to be opened to
	 * make it persistent.
	 */
	if ((fp = fopen(stfn, "r")) == NULL) {
		pnerror(errno, stfn);
		return 1;
	}
#endif
	while (wait(&status) != pid);
	t2 = times(&tp);
	sigset(SIGINT, oldint);
	sigset(SIGQUIT, oldquit);
	if (WIFSIGNALED(status))
		fprintf(stderr, "%s: command terminated abnormally.\n\n",
				progname);
	else
		fprintf(stderr, "\n");
#ifdef	PTIME
	if (ptimes(&tp, fp, stfn) < 0)
		return 1;
#endif
	out("real", t2 - t1);
	out("user", tp.tms_cutime);
	out("sys", tp.tms_cstime);
	return status ? (WIFEXITED(status)
			? WEXITSTATUS(status) : WTERMSIG(status) | 0200)
		: 0;
}

int
main(int argc, char **argv)
{
	int i;

	progname = basename(argv[0]);
	setlocale(LC_NUMERIC, "");
	localec = localeconv();
	i = 1;
	while (i < argc && argv[i][0] == '-' && argv[i][1]) {
	nxt:	switch (argv[i][1]) {
		case 'p':
			pflag = 1;
			if (argv[i][2] == 'p') {
				(argv[i])++;
				goto nxt;
			} else
				i++;
			break;
		case '-':
			if (argv[i][2] == '\0')
				i++;
			/*FALLTHRU*/
		default:
			goto opnd;
		}
	}
opnd:	if (i >= argc)
		return 0;
	hz = sysconf(_SC_CLK_TCK);
	return timecmd(&argv[i]);
}
